tests/unit/s2n_async_pkcs11_sign_test.c

```diff
@@ -0,0 +1,574 @@
+                                             NULL,
+                                             NULL, 
+                                             session));
+    free(slot_list);
+
+    return S2N_SUCCESS;
+}
+
+static int pkcs11_login(CK_SESSION_HANDLE session, CK_UTF8CHAR * pin, CK_ULONG pin_size)
+{
+    CK_FUNCTION_LIST_PTR function_list = NULL;
+    POSIX_GUARD(C_GetFunctionList(&function_list));
+    POSIX_GUARD_PTR(function_list);
+
+    POSIX_GUARD(function_list->C_Login(session,
+                                        CKU_USER,
+                                        pin,
+                                        pin_size));
+    return S2N_SUCCESS;
+}
+
+static int pkcs11_find_key(CK_SESSION_HANDLE session, const char * label, CK_ULONG label_size, CK_OBJECT_HANDLE_PTR key)
+{
+    CK_FUNCTION_LIST_PTR function_list = NULL;
+    POSIX_GUARD(C_GetFunctionList(&function_list));
+    POSIX_GUARD_PTR(function_list);
+
+    CK_ULONG count = 0;
+    CK_OBJECT_CLASS key_class = CKO_PRIVATE_KEY;
+
+    CK_ATTRIBUTE template[] = {
+        { .type = CKA_LABEL, 
+            .pValue = (CK_VOID_PTR) label, 
+            .ulValueLen = label_size
+        },
+        { .type = CKA_CLASS,
+            .pValue = &key_class,
+            .ulValueLen = sizeof(CK_OBJECT_CLASS),
+        }
+    };
+
+
+    POSIX_GUARD(function_list->C_FindObjectsInit(session, template, sizeof(template) / sizeof(CK_ATTRIBUTE)));
+
+    CK_OBJECT_HANDLE handle = CK_INVALID_HANDLE;
+    POSIX_GUARD(function_list->C_FindObjects(session,
+                                             &handle,
+                                             1UL,
+                                             &count));
+
+    POSIX_GUARD(function_list->C_FindObjectsFinal(session));
+    EXPECT_TRUE(handle != CK_INVALID_HANDLE);
+    EXPECT_TRUE(count != 0);
+
+    *key = handle;
+
+    return S2N_SUCCESS;
+}
+
+static int pkcs11_setup(CK_SESSION_HANDLE_PTR session, CK_OBJECT_HANDLE_PTR key, char * label, CK_ULONG label_size)
+{
+    POSIX_GUARD_PTR(session);
+    POSIX_GUARD_PTR(key);
+
+    POSIX_GUARD(pkcs11_init());
+    POSIX_GUARD(pkcs11_setup_session(session));
+    POSIX_GUARD(pkcs11_login(*session, (CK_UTF8CHAR_PTR)pkcs11_pin, sizeof(pkcs11_pin)-1UL));
+    POSIX_GUARD(pkcs11_find_key(*session, label, label_size, key));
+
+    return S2N_SUCCESS;
+}
+
+static int pkcs11_cleanup(CK_SESSION_HANDLE session)
+{
+    CK_FUNCTION_LIST_PTR function_list = NULL;
+    POSIX_GUARD(C_GetFunctionList(&function_list));
+    POSIX_GUARD_PTR(function_list);
+
+    POSIX_GUARD(function_list->C_CloseSession(session));
+    POSIX_GUARD(function_list->C_Finalize(NULL));
+
+    return S2N_SUCCESS;
+}
+
+
+static int pkcs11_decrypt(CK_SESSION_HANDLE session,
+                 CK_OBJECT_HANDLE key,
+                 const uint8_t * in, 
+                 uint32_t in_len,
+                 uint8_t ** out_buf, 
+                 uint32_t * out_len)
+{
+    POSIX_GUARD_PTR(in);
```

Comment:
Isn't this just a signing test? Do you need to implement decryption?

Reply:
We need the private key to decrypt messages for the functional test, otherwise the handshake will fail

Reply:
I'll take a stab at making the test more fine-grained in a separate PR as I chunk this work up